src/frontend/src/optimizer/plan_node/stream_sink.rs

```diff
@@ -315,6 +318,56 @@ impl StreamSink {
         let (pk, _) = derive_pk(input.clone(), user_order_by, &columns);
         let mut downstream_pk =
             Self::parse_downstream_pk(&columns, properties.get(DOWNSTREAM_PK_KEY))?;
+
+        {
+            // check for key encode TEXT
+            // if key encode is TEXT, we require the key to be a single column
+            // and the column should be of these types: varchar, bool, small int, int, big int
+            if let Some(format_desc) = &format_desc
+                && let Some(key_encode) = &format_desc.key_encode
+            {
+                if key_encode != &SinkEncode::Text {
+                    return Err(ErrorCode::SinkError(Box::new(Error::new(
+                        ErrorKind::InvalidInput,
+                        format!(
+                            "The key encode {:?} is not supported, only TEXT is supported.",
+                            &key_encode
+                        ),
+                    )))
+                    .into());
+                }
+                if downstream_pk.len() != 1 {
+                    return Err(ErrorCode::SinkError(Box::new(Error::new(
+                        ErrorKind::InvalidInput,
+                        format!(
+                            "The key encode is TEXT, but the primary key has {} columns. The key encode TEXT requires the primary key to be a single column.",
+                            downstream_pk.len()
+                        ),
+                    ))).into());
+                }
+
+                let key_col_catalog = columns.get(downstream_pk[0]).unwrap();
+                match &key_col_catalog.column_desc.data_type {
+                    DataType::Varchar
+                    | DataType::Boolean
+                    | DataType::Int16
+                    | DataType::Int32
+                    | DataType::Int64 => {}
+                    _ => {
+                        // why we don't allow float as text for key encode: https://github.com/risingwavelabs/risingwave/issues/6412
+                        return Err(ErrorCode::SinkError(Box::new(Error::new(
+                            ErrorKind::InvalidInput,
+                            format!(
+                                "The key encode is TEXT, but the primary key column {} has type {}. The key encode TEXT requires the primary key column to be of type varchar, bool, small int, int, or big int.",
+                                key_col_catalog.column_desc.name,
+                                key_col_catalog.column_desc.data_type
+                            ),
+                        ))).into());
```

Comment:
Just to confirm: these messages will be thrown to users during `CREATE SINK` command, right?

Reply:
yes. you can see the test case here

https://github.com/risingwavelabs/risingwave/blob/313216c9b5df27fb653b47db7a21fa4092ad2ddb/e2e_test/sink/kafka/create_sink.slt#L208-L216